service/storage_service.cc

```diff
@@ -502,7 +502,7 @@ future<storage_service::nodes_to_notify_after_sync> storage_service::sync_raft_t
 
         switch (rs.state) {
         case node_state::bootstrapping:
-            if (!rs.ring.tokens.empty()) {
+            if (!rs.ring.tokens.empty()) { // I'm not sure if not executing this code for zero-token nodes is right. Maybe we skip something important. I need to investigate it.
```

Comment:
I think this is used to check which stage of bootstrap process we're in (for normal node: before or after assigning tokens). So if it turns out to be important (once you investigate) maybe we can check the stage directly from state machine instead of looking at tokens

Reply:
`update_topology` call might be important.
not sure about:
```
                        tmptr->add_bootstrap_tokens(rs.ring.tokens, host_id);
                        co_await update_topology_change_info(tmptr, ::format("bootstrapping node {}/{}", id, ip));
```
since the node has no tokens and we skip streaming.

Reply:
```cpp
case node_state::bootstrapping:
    if (!rs.ring.tokens.empty()) {
```

We can enter this code only after assigning tokens to a joining node, which happens in the `join_group0` state. For joining zero-token nodes, this is the last state. So, instead of executing this code, we simply execute `process_normal_node` for zero-token nodes after the `join_group0` state.

> `update_topology` call might be important. not sure about:

It is called in `process_normal_node`.

> not sure about:
> ```
>                         tmptr->add_bootstrap_tokens(rs.ring.tokens, host_id);
>                         co_await update_topology_change_info(tmptr, ::format("bootstrapping node {}/{}", id, ip));
> ```
> 
> since the node has no tokens and we skip streaming.

We should not do this for zero-token nodes or this should be a no-op for zero-token nodes. The idea is that zero-token nodes are normal nodes from the topology point of view, but they are not existent from the ring and tablets point of view. I'll check that zero-token nodes do not wander unnoticed somewhere in the `token_metadata` logic just to be sure.

Reply:
So, zero-token nodes appear in `token_metadata_impl::replacing_endpoints` and `token_metadata_impl::_leaving_endpoints`. This is completely harmless. I could try to change it, but I doubt it's worth the effort.